### PR TITLE
[fix/improve_photo_upload_responsiveness] More reliable photo upload

### PR DESCRIPTION
**Pull request by mneuwert**
## Description
It is addressing the below issue but actually fixing even more issues including:
- Media files not being imported when app crashes / being terminated by the iOS
- Resource bottle-necks when uploading big amounts of media files
- Reduces number of sync operations when uploading lots of media files

## Related Issue
#504 

## Motivation and Context
See description and related issue

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### QA

Test plan:

https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%201.2.0/Uploads%20Responsiveness.md

- [X] Stucked when importing [FIXED] https://github.com/owncloud/ios-app/pull/548#issuecomment-560377289

### CODE REVIEW

File: ownCloud/UIKit Extensions/UIViewController+Extension.swift [hunk 1]
```diff
@@ -34,4 +34,21 @@ extension UIViewController {
 			tabBarController.toolbar?.setItems(nil, animated: true)
 		}
 	}
+
+	func topMostViewController() -> UIViewController {
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/UIKit Extensions/UIViewController+Extension.swift]: @felix-schwarz fixed

File: ownCloud/Client/Viewer/Media/MediaDisplayViewController.swift [hunk 2]
```diff
@@ -31,11 +31,16 @@ class MediaDisplayViewController : DisplayViewController {
 	deinit {
 		playerStatusObservation?.invalidate()
 		playerItemStatusObservation?.invalidate()
+		NotificationCenter.default.removeObserver(self)
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/Client/Viewer/Media/MediaDisplayViewController.swift]: @felix-schwarz fixed

File: ownCloud/PhotoKit Extensions/MediaUploadQueue.swift [hunk 3]
```diff
@@ -23,91 +23,141 @@ import MobileCoreServices
...
-			})
+		vault.keyValueStore?.storeObject(uploads as NSDictionary, forKey: MediaUploadQueue.PendingAssetsKey)
+	}
+
+	class func removePending(asset identifier:String, for bookmark:OCBookmark) {
+		let vault : OCVault = OCVault(bookmark: bookmark)
+		var uploads = vault.keyValueStore?.readObject(forKey: MediaUploadQueue.PendingAssetsKey) as? [String : String]
+		if uploads != nil {
+			uploads?.removeValue(forKey: identifier)
+			vault.keyValueStore?.storeObject(uploads! as NSDictionary, forKey: MediaUploadQueue.PendingAssetsKey)
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/PhotoKit Extensions/MediaUploadQueue.swift]: @felix-schwarz fixed

File: ownCloud/PhotoKit Extensions/MediaUploadQueue.swift [hunk 4]
```diff
@@ -23,91 +23,141 @@ import MobileCoreServices
...
 
-			uploadGroup.notify(queue: queue, execute: {
-				if weakCore != nil {
-					MediaUploadQueue.resetUploadPendingFlag(for:  weakCore!.bookmark)
-				}
-				MediaUploadQueue.uploadStarted = false
-				backgroundTask?.end()
-				assetUploadCompletion?(nil, true)
-			})
+		vault.keyValueStore?.storeObject(uploads as NSDictionary, forKey: MediaUploadQueue.PendingAssetsKey)
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/PhotoKit Extensions/MediaUploadQueue.swift]: @felix-schwarz fixed

File: ownCloud/AppDelegate.swift [hunk 5]
```diff
@@ -92,9 +95,39 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
 
+		// Subscribe to media upload queue notifications
+		NotificationCenter.default.addObserver(self, selector: #selector(handleAssetImportStarted(notification:)), name: MediaUploadQueue.AssetImportStarted.name, object: nil)
+		NotificationCenter.default.addObserver(self, selector: #selector(handleAssetImportFinished), name: MediaUploadQueue.AssetImportFinished.name, object: nil)
+		NotificationCenter.default.addObserver(self, selector: #selector(handleSingleAssetImport), name: MediaUploadQueue.AssetImported.name, object: nil)
+
 		return true
 	}
 
+	private func updateImportMediaHUD() {
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/AppDelegate.swift]: @felix-schwarz fixed

File: ownCloud/AppDelegate.swift [hunk 6]
```diff
@@ -92,9 +95,39 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 		UIApplication.shared.setMinimumBackgroundFetchInterval(
 			UIApplication.backgroundFetchIntervalMinimum)
 
+		// Subscribe to media upload queue notifications
+		NotificationCenter.default.addObserver(self, selector: #selector(handleAssetImportStarted(notification:)), name: MediaUploadQueue.AssetImportStarted.name, object: nil)
+		NotificationCenter.default.addObserver(self, selector: #selector(handleAssetImportFinished), name: MediaUploadQueue.AssetImportFinished.name, object: nil)
+		NotificationCenter.default.addObserver(self, selector: #selector(handleSingleAssetImport), name: MediaUploadQueue.AssetImported.name, object: nil)
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/AppDelegate.swift]: @felix-schwarz fixed

File: ownCloud/AppDelegate.swift [hunk 7]
```diff
@@ -24,6 +24,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
 	var window: UIWindow?
 	var serverListTableViewController: ServerListTableViewController?
+	var hud: ProgressHUDViewController?
+	var importedMediaCount: NSNumber?
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/AppDelegate.swift]: @felix-schwarz fixed

File: ownCloud/AVFoundation Extensions/AVAsset+Extension.swift [hunk 8]
```diff
@@ -19,28 +19,34 @@
...
-				completion(false)
-				return
+			if compatiblePreset {
+				guard let export = AVAssetExportSession(asset: self, presetName: preset) else {
+					return false
+				}
+				export.outputFileType = type
+				export.outputURL = targetURL
+				export.exportAsynchronously {
+					exportSuccess = (export.status == .completed)
```
[the remark this answers is not in the document]
Comment (mneuwert) [reply ownCloud/AVFoundation Extensions/AVAsset+Extension.swift]: @felix-schwarz indeed, good catch.. fixed

**Comment by mneuwert** [comment]
@felix-schwarz I think this is now ready for another code review after refactoring. Can you please have a look?

**Comment by jesmrec** [comment]
First QA findings:

About manual uploads from Photo albums

Uploads run faster, but importing keeps having bad UX. By importing ~30-50 files, you have to wait a bit till it starts. If you select to import +150 files, the app is stucked till it finishes with no indicator for the user that something is happening (like description here: https://github.com/owncloud/ios-app/issues/504#issue-497040012). Sometimes, finally crashes, other times, after waiting for more that 2 minutes, the uploads start. A regular user might kill the app at that moment.  

iPhoneX 13.2.2
oC10.3

(keep working on it)

**Comment by jesmrec** [comment]
Problem fixed. Now, app is not stucked when a big amount of files is imported.

**Pull request merged**